core/trino-main/src/main/java/io/trino/sql/planner/CaseExpressionPredicateRewriter.java

```diff
@@ -0,0 +1,279 @@
+        {
+            return caseExpression.getWhenClauses().stream()
+                    .map(x -> ((ComparisonExpression) x.getOperand()).getLeft())
+                    .findFirst().get();
+        }
+
+        private boolean canRewriteSearchedCaseExpression(SearchedCaseExpression caseExpression)
+        {
+            ImmutableList.Builder<Expression> rightHandSideExpressions = ImmutableList.builder();
+            ImmutableList.Builder<Expression> leftHandSideExpressions = ImmutableList.builder();
+            for (WhenClause whenClause : caseExpression.getWhenClauses()) {
+                Expression whenOperand = whenClause.getOperand();
+                if (!(whenOperand instanceof ComparisonExpression)) {
+                    return false;
+                }
+                ComparisonExpression whenComparisonFunction = (ComparisonExpression) whenOperand;
+                Expression left = whenComparisonFunction.getLeft();
+                Expression right = whenComparisonFunction.getRight();
+
+                if (!whenComparisonFunction.getOperator().equals(EQUAL)) {
+                    return false;
+                }
+                leftHandSideExpressions.add(left);
+                rightHandSideExpressions.add(right);
+            }
+            return checkSimilarAndDeterministic(leftHandSideExpressions.build()) &&
+                    checkNonNullUniqueLiterals(rightHandSideExpressions.build());
+        }
+
+        private boolean checkSimilarAndDeterministic(List<Expression> expressions)
+        {
+            Expression first = null;
+            for (Expression expression : expressions) {
+                if (first == null) {
+                    first = expression;
+                    if (!DeterminismEvaluator.isDeterministic(expression, plannerContext.getMetadata())) {
+                        return false;
+                    }
+                }
+                if (!first.equals(expression)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private boolean checkNonNullUniqueLiterals(List<Expression> expressions)
+        {
+            Set<Object> literals = new HashSet<>();
+            Type type = null;
+            for (Expression expression : expressions) {
+                if (expression instanceof NullLiteral || !isEffectivelyLiteral(plannerContext, session, expression)) {
+                    return false;
+                }
+                if (type == null) {
+                    type = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression).get(NodeRef.of(expression));
+                }
+                Object constantExpression = evaluateConstantExpression(expression, type, plannerContext, session, new AllowAllAccessControl(), ImmutableMap.of());
+                if (constantExpression == null || constantExpression instanceof NullLiteral || literals.contains(constantExpression)) {
```

Comment:
`evaluateConstantExpression` never returns an `Expression`, so you can remove `|| constantExpression instanceof NullLiteral`.

Reply:
Done.